File: Data Structures & Algorithms/maximum-product-subarray/submission-6.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int max_prod = 1, min_prod = 1;
        int result = nums[0];
        for (int n : nums) {
            int temp_max = max_prod;
            max_prod = max({n, max_prod * n, min_prod * n});
            min_prod = min({n, temp_max * n, min_prod * n});
            result = max(result, max_prod);
        }
        return result;
    }
};

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-6.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        result = [1] * len(nums)
        for i in range(1, len(nums)):
            for j in range(i):
                if nums[j] >= nums[i]:
                    continue
                result[i] = max(result[i], 1 + result[j])
        return max(result)